### PR TITLE
Outfit lua metatable

### PR DESCRIPTION
**Pull request by nekiro**
This pr adds Outfit() metatable to use in lua. Helpful in cases when you for example need to get name of outfit from looktype.

Outfit(looktype)
and outfit compare method

### CODE REVIEW

**Review by Znote — APPROVED** [review]
Beautiful and educational code that works great as a sample on how to metatablify something. :)

File: src/outfit.cpp [hunk 1]
```diff
@@ -75,3 +75,15 @@ const Outfit* Outfits::getOutfitByLookType(PlayerSex_t sex, uint16_t lookType) c
 	}
 	return nullptr;
 }
+
+const Outfit* Outfits::getOutfitByLookType(uint16_t lookType) const
+{
+	for (uint8_t sex = 0; sex <= 1; sex++) {
```
Comment (marmichalski) [inline src/outfit.cpp:81]: There are enums for genders + `PLAYERSEX_LAST`.

File: src/outfit.h [hunk 2]
```diff
@@ -26,6 +26,11 @@ struct Outfit {
 	Outfit(std::string name, uint16_t lookType, bool premium, bool unlocked) :
 		name(std::move(name)), lookType(lookType), premium(premium), unlocked(unlocked) {}
 
+	bool operator==(const Outfit& otherOutfit) const
+	{
+		return (name == otherOutfit.name) && (lookType == otherOutfit.lookType) && (premium == otherOutfit.premium) && (unlocked == otherOutfit.unlocked);
```
Comment (marmichalski) [inline src/outfit.h:31]: All those parentheses are redundant.

**Comment by nekiro** [comment]
Thanks  @marmichalski, not sure why this pr had so much of weird code, but it was pretty old.

**Comment by ranisalt** [comment]
> Beautiful and educational code that works great as a sample on how to metatablify something. :)

I think @nekiro 's thought process deserves an entry in the wiki ;)

**Comment by nekiro** [comment]
> > Beautiful and educational code that works great as a sample on how to metatablify something. :)
> 
> I think @nekiro 's thought process deserves an entry in the wiki ;)

Are you rude???

**Comment by Znote** [comment]
This did not work as good as I was hoping for, got time to test it out today:

```lua
local talkaction = TalkAction("!hello")

function talkaction.onSay(player)
   print(player:getName() .. ' executed talkaction: !hello')
   
   tdump("player:getOutfit()", player:getOutfit())
   tdump("Outfit(140)", Outfit(140))
   tdump("Outfit(140) == Outfit(140) -- Noblewoman == Noblewoman", Outfit(140) == Outfit(140))
   tdump("Outfit(140) == player:getOutfit() -- Noblewoman == Noblewoman", Outfit(140) == player:getOutfit())
   tdump("Outfit(140) == Outfit(151) -- Noblewoman == Pirate", Outfit(140) == Outfit(151))
   tdump("Outfit(140) == Outfit(132) -- Noblewoman == Nobleman", Outfit(140) == Outfit(132))

   return true
end

talkaction:register()
```

image

`player:getOutfit()` is unable to compare against `Outfit()`, even when looktype matches. 
`player:getOutfit()` is missing data such as premium, unlocked, name

Is there no way to compare "matching set across gender"? I want `Outfit(132) == Outfit(140)` to match true as in "I want to see if this is noble outfit" regardless of gender (and thus looktype). 

I find it a bit redundant to first get player outfit, then initialize a new outfit object before I can compare it. 
```lua
local noblewoman = Outfit(140)
if noblewoman ~= Outfit(player:getOutfit().lookType) then
    -- Only fancy rich women allowed!
end
```

**Comment by ranisalt** [comment]
@Znote but an outfit is more than just the looktype, there are colors, addons, gender and more as you shown. How do you want two outfits to equal when the only value that is equal is the looktype? It should read `if player:getOutfit().lookType == 140 then` if you want to check if that's the noble[wo]man outfit.

**Comment by Znote** [comment]
I argue that since the object is created from lookType, it should be able to match against player:getOutfit(), it doesn't matter which colours you dress, your either wearing noblewomen or your not. I would rather take an extra step in the code if the color mattered, instead of vice versa. 

But even this:
```lua
if player:getOutfit().lookType == Outfit("Noblwomen").lookType then
    -- ...
end
```
Would be a nice improvement, I doubt everyone remembers what lookType 140 is, but if its initialized as "Noblewomen" then its pretty easy to understand what outfit we are working with, what outfit we want the player to have. 

Yeah I figured the gender match was a stretch, the XML file at least has to be structured differently to accomodate gender groupings of outfits.

**Comment by marmichalski** [comment]
@Znote you're mixing something I'd call PlayerOutfit with XmlOutfit.

**Comment by Kamenuvol** [comment]
Maybe call OutfitType metatable instead

**Pull request merged**